pkg/clientaccess/token_test.go

```diff
@@ -0,0 +1,318 @@
+package clientaccess
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/rancher/dynamiclistener/cert"
+	"github.com/rancher/dynamiclistener/factory"
+	"github.com/rancher/k3s/pkg/bootstrap"
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	defaultUsername = "server"
+	defaultPassword = "token"
+)
+
+// TestTrustedCA confirms that tokens are validated when the server uses a cert (self-signed or otherwise)
+// that is trusted by the OS CA bundle. This test must be run first, since it mucks with the system root certs.
+func TestTrustedCA(t *testing.T) {
+	assert := assert.New(t)
+	server := newTLSServer(t, defaultUsername, defaultPassword)
+	defer server.Close()
+
+	testInfo := &Info{
+		CACerts:  getServerCA(server),
+		BaseURL:  server.URL,
+		Username: defaultUsername,
+		Password: defaultPassword,
+		caHash:   hashCA(getServerCA(server)),
+	}
+
+	testCases := []struct {
+		token    string
+		expected string
+	}{
+		{defaultPassword, ""},
+		{testInfo.String(), testInfo.Username},
+	}
+
+	// Point OS CA bundle at this test's CA cert to simulate a trusted CA cert.
+	// Note that this only works if the OS CA bundle has not yet been loaded in this process,
+	// as it is cached for the duration of the process lifetime.
+	// Ref: https://github.com/golang/go/issues/41888
+	path := t.TempDir() + "/ca.crt"
```

Comment:
The `t.TempDir()` is available at go >= 1.15 https://golang.org/doc/go1.15#testing

Reply:
Yes, we're using 1.15. Why would ioutils version be preferred when the one I'm using here is automatically cleaned up at the end of the test?

Reply:
After

> Yes, we're using 1.15. Why would ioutils version be preferred when the one I'm using here is automatically cleaned up at the end of the test?

The purpose is to make k3s supports go <= 1.14, but I found the project using go 1.15.x.
NVM, I'm going to upgrade my local go version.

Reply:
Yeah, some of the branches are still on 1.13, while newer ones are on 1.15. We shouldnt have anything on 1.14 due to issues with signal handling.